cmd/ib-sriov-cni/main.go

```diff
@@ -215,59 +248,65 @@ func runIPAMPlugin(stdinData []byte, netConf *localtypes.NetConf) (_ *current.Re
-	netConf, netns, err := getNetConfNetns(args)
+// handleVfioPciMode handles VFIO PCI devices with VF GUID setting if needed
+func handleVfioPciMode(args *skel.CmdArgs, netConf *localtypes.NetConf, netns ns.NetNS) error {
+	// Check if the device is a VF (Virtual Function) or PF (Physical Function)
+	isVF, err := utils.IsVirtualFunction(netConf.DeviceID)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to determine if device %s is VF or PF: %v", netConf.DeviceID, err)
 	}
-	defer netns.Close()
 
-	sm := sriov.NewSriovManager()
+	// Only handle GUID setting for VF devices, keep previous code path for PF devices
+	if isVF && netConf.GUID != "" {
+		// Load device info needed for GUID setting (VFIO VF version - no network interface)
+		err = config.LoadDeviceInfoVfioVF(netConf)
+		if err != nil {
+			return fmt.Errorf("failed to get device specific information for vfio-pci VF device: %v", err)
+		}
 
-	// Lock CNI operation to serialize the operation
-	lock, err := lockCNIExecution()
-	if err != nil {
-		return err
+		sm := sriov.NewSriovManager()
+		if err := sm.ApplyVFConfig(netConf); err != nil {
+			return fmt.Errorf("failed to configure VF GUID for vfio-pci device: %v", err)
+		}
 	}
-	defer unlockCNIExecution(lock)
 
-	err = doVFConfig(sm, netConf, netns, args)
-	if err != nil {
-		return err
+	result := &current.Result{}
+	result.Interfaces = []*current.Interface{{
+		Name:    args.IfName,
+		Sandbox: netns.Path(),
+	}}
+
+	// Cache NetConf for CmdDel (minimal config for vfio-pci mode)
+	if err = utils.SaveNetConf(args.ContainerID, config.DefaultCNIDir, args.IfName, netConf); err != nil {
+		return fmt.Errorf("error saving NetConf %q", err)
 	}
-	defer func() {
-		if retErr != nil {
-			nsErr := netns.Do(func(_ ns.NetNS) error {
-				_, innerErr := netlink.LinkByName(args.IfName)
-				return innerErr
-			})
-			if nsErr == nil {
-				_ = sm.ReleaseVF(netConf, args.IfName, args.ContainerID, netns)
-			}
-			if netConf.RdmaIso {
-				_ = utils.MoveRdmaDevFromNs(netConf.RdmaNetState.ContainerRdmaDevName, netns)
-			}
-		}
-	}()
 
+	return types.PrintResult(result, netConf.CNIVersion)
+}
+
+// setupIPAMAndResult handles IPAM configuration and creates the result
+func setupIPAMAndResult(args *skel.CmdArgs, netConf *localtypes.NetConf, netns ns.NetNS) (*current.Result, func(error), error) {
 	result := &current.Result{}
 	result.Interfaces = []*current.Interface{{
 		Name:    args.IfName,
 		Sandbox: netns.Path(),
 	}}
 
+	// Default cleanup function (no-op)
+	cleanup := func(error) {}
+
 	if netConf.IPAM.Type != "" {
-		var newResult *current.Result
-		newResult, err = runIPAMPlugin(args.StdinData, netConf)
+		newResult, err := runIPAMPlugin(args.StdinData, netConf)
 		if err != nil {
-			return err
+			return nil, cleanup, err
 		}
-		// If runIPAMPlugin failed, than ExecDel was called. Defer if no error
-		defer func() {
+
+		// Return cleanup function for IPAM
+		cleanup = func(retErr error) {
```

Comment:
iMO cleanup func should not accept an error, but rather just execute `ipam.ExecDel` the caller can then wrap it with whatever checks are needed (e.g check retErr).

so cleanup should just be `func()`

Reply:
Agreed. With your suggested improvements and fixes in the latest code, this function is no longer needed. I’ll keep the code changes minimal and focus on supporting VFIO only in the next version.